torch/csrc/autograd/profiler.cpp

```diff
@@ -93,27 +93,126 @@ void popRange() {
   }
 }
 
-RecordFunction::RecordFunction(Function* fn) {
-  // typeid(*fn).name() would avoid an additional string allocation.
-  // However, typeid(*fn).name() would cause nvtx annotations for all user-defined
-  // (Python-side) custom autograd function backward() methods to have the same name,
-  // because they route through the same C++ side class.
-  // fn->name() ensures that nvtx annotations for custom function backward() methods
-  // receive a relevant, demangled name.
+// typeid(*fn).name() would avoid an additional string allocation.
+// However, typeid(*fn).name() would cause nvtx annotations for all user-defined
+// (Python-side) custom autograd function backward() methods to have the same name,
+// because they route through the same C++ side class.
+// fn->name() ensures that nvtx annotations for custom function backward() methods
+// receive a relevant, demangled name.
+FunctionCallContext::FunctionCallContext(Function* fn)
+    : fn_(fn), owned_name_(new std::string(fn->name())), name_ptr_(owned_name_->c_str()),
+      sequence_nr_(fn->sequence_nr()) {}
+
+FunctionCallContext::FunctionCallContext(
+    std::string name, int64_t sequence_nr)
+    : owned_name_(new std::string(std::move(name))), name_ptr_(owned_name_->c_str()),
+      sequence_nr_(sequence_nr) {}
+
+FunctionCallContext::FunctionCallContext(
+    const char* name_ptr, int64_t sequence_nr)
+    : name_ptr_(name_ptr), sequence_nr_(sequence_nr) {}
+
+
+namespace {
+std::mutex callbacks_mutex;
+std::atomic<RecordCallbacks*> global_callbacks(nullptr);
+}
+
+void registerCallbacks(
+    const RecordFunctionCallback& start, const RecordFunctionCallback& end) {
+  std::lock_guard<std::mutex> guard(callbacks_mutex);
+  RecordCallbacks* new_callbacks = new RecordCallbacks;
+  RecordCallbacks* prev_callbacks = global_callbacks;
+  if (prev_callbacks) {
+    *new_callbacks = *prev_callbacks;
+  }
+  new_callbacks->start_callbacks.push_back(start);
+  new_callbacks->end_callbacks.push_back(end);
+  global_callbacks = new_callbacks;
+  if (prev_callbacks) {
+    delete prev_callbacks;
+  }
+}
+
+void clearCallbacks() {
+  std::lock_guard<std::mutex> guard(callbacks_mutex);
+  RecordCallbacks* prev_callbacks = global_callbacks;
+  global_callbacks = nullptr;
+  if (prev_callbacks) {
+    delete prev_callbacks;
+  }
+}
+
+RecordFunction::RecordFunction(
+    Function* fn, UpdateCallContextCallback* ucb)
+    : callbacks_(global_callbacks) {
   pushRangeImpl(fn->name(), ", stashed seq=", fn->sequence_nr());
+  if (callbacks_) {
```

Comment:
We never check that the profiler is actually on here? That seems weird

Reply:
we don't necessarily want to check profiler flag for the callbacks